lib/migration/src/Obelisk/Migration.hs

```diff
@@ -0,0 +1,112 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+module Obelisk.Migration where
+
+import Control.Monad (forM)
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Maybe (catMaybes, listToMaybe)
+import Data.Semigroup ((<>))
+import qualified Data.Set as Set
+import Data.Text (Text)
+import qualified Data.Text as T
+import System.Directory (doesDirectoryExist, doesFileExist, listDirectory)
+import System.FilePath
+
+-- TODO: Add error handling in place of error/fail functions
+
+import Algebra.Graph.AdjacencyMap
+
+type Hash = Text
+
+-- Migration graph with edge metadata
+data Migration action = Migration
+  { _migration_graph :: AdjacencyMap Hash
+  , _migration_actions :: Map (Hash, Hash) action -- Edge metadata
+  }
+  deriving (Eq, Show)
+
+-- | Get the directed-acylic graph of migration
+getDag :: Ord a => AdjacencyMap a -> AdjacencyMap a
+getDag g = case topSort g of
+  Just _ -> g
+  Nothing -> error "Cyclic graph detected"
+
+-- Return actions that migrate from one hash to another
+runMigration :: Migration action -> Hash -> Hash -> Maybe [(Hash, action)]
+runMigration m a b = fmap getAction <$> findPath a b m
```

Comment:
All paths in the DAG from `a` to `b` must sum to the same `action` value, otherwise the migration is invalid.  It looks like that check may still need to be added.

Reply:
Good point; that would also tell us the 'integrity' of the migration graph.

Reply:
Addressed this in 259b853 and 2e81b11